atintegrators/BeamMomentsPass.c

```diff
@@ -15,79 +15,71 @@
 struct elem
 {
   int turn;
-  double *sizes;
-  double *positions;
+  double *stds;
+  double *means;
 };
 
 
 void BeamMomentsPass(double *r_in, int nbunch, int num_particles, struct elem *Elem) {
 
     int turn = Elem->turn;
-    double *sizes = Elem->sizes;
-    double *positions = Elem->positions;
+    double *stds = Elem->stds;
+    double *means = Elem->means;
         
     int i, ii, ib; 
-    size_t sz = 2*nbunch*6*sizeof(double) + nbunch*sizeof(double);
-    void *buffer = atMalloc(sz);
+    void *buffer = atCalloc(2*nbunch*6+nbunch, sizeof(double));
     double *dptr = (double *) buffer;
     
-    double *avep=dptr; dptr += nbunch*6;
-    double *sizep=dptr; dptr += nbunch*6;
+    double *meanp=dptr; dptr += nbunch*6;
+    double *stdp=dptr; dptr += nbunch*6;
     double *nparts=dptr;
     
-    for (i=0; i<nbunch; i++) {
-        nparts[i] = 0.0;
-        for(ii=0; ii<6; ii++) {
-            avep[i*6+ii] = 0.0;
-            sizep[i*6+ii] = 0.0;
-        }
-    }
-    
     for (i=0; i<num_particles; i++) {
         double *r6 = r_in+i*6;
         if(!atIsNaN(r6[0])){
             ib = i%nbunch;
             nparts[ib] += 1;
             for(ii=0; ii<6; ii++) {
-                avep[6*ib+ii] += r6[ii];
-                sizep[6*ib+ii] += r6[ii]*r6[ii];
+                meanp[6*ib+ii] += r6[ii];
+                stdp[6*ib+ii] += r6[ii]*r6[ii];
             }
         }
     }   
     
     #ifdef MPI     
-    MPI_Allreduce(MPI_IN_PLACE,sizep,6*nbunch,MPI_DOUBLE,MPI_SUM,MPI_COMM_WORLD);
-    MPI_Allreduce(MPI_IN_PLACE,avep,6*nbunch,MPI_DOUBLE,MPI_SUM,MPI_COMM_WORLD);
+    MPI_Allreduce(MPI_IN_PLACE,stdp,6*nbunch,MPI_DOUBLE,MPI_SUM,MPI_COMM_WORLD);
+    MPI_Allreduce(MPI_IN_PLACE,meanp,6*nbunch,MPI_DOUBLE,MPI_SUM,MPI_COMM_WORLD);
     MPI_Allreduce(MPI_IN_PLACE,nparts,nbunch,MPI_DOUBLE,MPI_SUM,MPI_COMM_WORLD);
     MPI_Barrier(MPI_COMM_WORLD);
     #endif
     
     for (i=0; i<nbunch; i++){       
         for (ii=0; ii<6; ii++){
-            avep[6*i+ii] = avep[6*i+ii]/nparts[i];
-            sizep[6*i+ii] = sqrt(sizep[6*i+ii]/nparts[i]-avep[6*i+ii]*avep[6*i+ii]); 
+            meanp[6*i+ii] = meanp[6*i+ii]/nparts[i];
+            stdp[6*i+ii] = sqrt(stdp[6*i+ii]/nparts[i]-meanp[6*i+ii]*meanp[6*i+ii]); 
         }
     }    
      
-    positions += 6*nbunch*turn;
-    sizes += 6*nbunch*turn;
-    memcpy(positions, avep, 6*nbunch*sizeof(double)); 
-    memcpy(sizes, sizep, 6*nbunch*sizeof(double));     
+    means += 6*nbunch*turn;
+    stds += 6*nbunch*turn;
+    memcpy(means, meanp, 6*nbunch*sizeof(double)); 
+    memcpy(stds, stdp, 6*nbunch*sizeof(double));
+    free(buffer);  
```

Comment:
done

Reply:
> You must use `atFree`. In Matlab, we are using the Matlab memory allocator, so it must be consistent with `atCalloc`. In python, we are using the standard C allocator, so it would work anyway.

@lfarv, should I always use `atMalloc` /`atFree`? In several places `malloc`/`free` was used instead...

Reply:
`malloc`/`free` always works: it's standard C. But Mathworks claims that its allocator is better, that's why there are the macros `atMalloc`/`AtFree` which do the conversion to the proper low-level allocator. It's preferable to use them, but the critical point is that you must not mix them: allocating with one and freeing with another one will crash or corrupt the memory.

Reply:
Understood, thanks